### DISCUSSION TITLE
128 KB program in AVR128DB flash?

### DISCUSSION THREAD

**Discussion by killerkalamari** [SpenceKonde/DxCore#137] [Q&A] [answered]
Can I write a program up to 128 KB or is the maximum program size still only 32 KB with the rest reserved for PROGMEM? If I can write more than 32 KB, does the compiler handle the details of that for me behind the scenes, or will I need to manually break up the code into 32 KB banks somehow? If the code can be larger, what happens if the currently executing code is at, say, offset 100 KB into the flash, but tries to access PROGMEM at offset 40 KB? Or, what happens if it needs to branch from flash 110 KB to 50 KB?

**Comment by killerkalamari** [comment, marked as answer]
Okay, it looks like the answers are:

- The program should be able to be the full 128 KB (Code Space: 0x00000 - 0x1FFFF), not just 32 KB. No banks needed, because the JMP instruction handles 22 bits of address (more than needed).
- Z=RAMPZ:R31:R30 with the ELPM instruction gives 24 addressable bits, so PROGMEM should be able to be used from anywhere, and access anywhere in the flash.

**Comment by SpenceKonde** [comment]
This is correct, instructions to be executed are read directly from flash and the 16 bit program counter allows for flash up to 128k because it;s word addressed.. The 32k segment limit only applies to reading from progmem using LD/ST, (mapped progmem) from the application.(we have a hacky trick to put variables in there described in the readme somewhere, and you can then read them like they were normal variables (though not write them, generally). Note that I don't think you save any *time* with this - I think it's more a matter of convenience and gracefulness) LD is 1 clock faster than LPM, but if you read the instructionset manual they say a minimum of 1 extra clock is needed to address stuff in mapped progmem, and refer to the datasheet for specifics, but I've read the relevant sections many times looking for the documentation of that and found nothing. So I think only the 1 clock minimum added execution time applies.

**Reply by killerkalamari** [reply]
The main advantage I see to using PROGMEM is RAM savings for large arrays/structs of predefined data, where without PROGMEM, the RAM would be pre-allocated and may be insufficient.

**Reply by SpenceKonde** [reply]
Well yeah, but you don't need to use the memory mapped segment to access variables in progmem, it's just more convenient, since you access like any normal variable (just slightly slower >.>) instead of with the pgm_read_datatype_near() macros.

**Reply by SpenceKonde** [reply]
(sideote: 
Yeah, it looks like 1 clock added to LD instructions that gothrough the NVM . So the only benefit is that it's handled in straight and the compiler can help you out moire (like most libraries in that location,pgmspace just wrappers  around instructions that do things C doesn't understand (like multiple address spaces  or watchdog tirmers ior sleep modes. 
On AVR64Dx, you can access all of the flash without having to use farptrs and elpm. and half also is mapped to the same address spoace as ram. On the AVR128Dx you can use progrem on the low half , and even without changing flmap nor using ELPM, access 3/4ths of the flash so I don't think there's any shortage of accaessible flashITalso writable from the app if you enable that option :-P via the included Flash library.
